ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/DistributingExecutor.java

```diff
@@ -241,6 +251,64 @@ public StatementExecutorResponse execute(
     }
   }
 
+  /**
+   * @return a list of warning messages for deprecated syntax statements
+   */
+  private List<KsqlWarning> getDeprecatedWarnings(
+      final MetaStore metaStore,
+      final ConfiguredStatement<?> statement
+  ) {
+    final List<KsqlWarning> deprecatedWarnings = new ArrayList<>();
+
+    if (isLeftOrOuterStreamStreamJoinWithoutGrace(metaStore, statement.getStatement())) {
+      deprecatedWarnings.add(new KsqlWarning(
+          "DEPRECATION NOTICE: Left/Outer stream-stream joins statements without a GRACE PERIOD "
+              + "will not be accepted in a future ksqlDB version.\n"
+              + "Please use the GRACE PERIOD clause as specified in "
+              + "https://docs.ksqldb.io/en/latest/developer-guide/ksqldb-reference/"
+              + "select-push-query/"
+      ));
+    }
+
+    return deprecatedWarnings;
+  }
+
+  private boolean isLeftOrOuterStreamStreamJoinWithoutGrace(
+      final MetaStore metaStore,
+      final Statement statement
+  ) {
+    // Check only CREATE_AS statements.
+    // Pull and push queries are not written in the command topic.
```

Comment:
Not sure if I can follow? While I understand that we cannot send back a warning for push queried right now (because we use web sockets), I don't see the relationship to the command topic?

Also, if we deprecate something else, we should also be able to send a warning back for pull queries (and after we moved off web sockets also for push queries)?

Reply:
Sorry, I didn't explain very well in that comment. The `DistributingExecutor` is the one that distributes commands through the command topic; waits for an answer; then responds back to the client. Only DDL commands are distributed because they change the metastore state of every node. Push and pull queries do not follow that path, so the reason this code only handles CREATE_AS operations. I think it was lazy to write all the above lines. Just wanted to mention this code path does not handle push/pull queries and it won't as that belongs somewhere else.

Reply:
What I am still wondering: why can not have a single place in the code that would add warning for all query types? Seems, if we add the code here, we would need to duplicate it elsewhere in the future?

Reply:
Done. While fixing the n-way joins, I moved the code to a different class `DeprecatedStatementsChecker` that checks if a statement is deprecated or not. We can reuse this class for push/pull queries later.